concreteproperties/concrete_section.py

```diff
@@ -1036,186 +1036,215 @@ def calculate_ultimate_section_actions(
+            and all ``control_points``. The length of ``labels`` must equal ``1`` or
+            ``2 + len(control_points)``.
+        :param n_points: Number of points to compute including and between the
+            ``limits`` of the moment interaction diagram. Generates equally spaced
+            neutral axes between the ``limits``.
+        :param n_spacing: If provided, overrides ``n_points`` and spaces the points on
+            the moment interaction diagram by an axial force ``n_spacing``. Note that
+            using ``n_spacing`` negatively affects performance as the neutral axis depth
+            must be located for each point on the moment interaction diagram.
         :param max_comp: If provided, limits the maximum compressive force in the moment
             interaction diagram to ``max_comp``
         :param max_comp_labels: Labels to apply to the ``max_comp`` intersection points,
             first value is at zero moment, second value is at the intersection with the
             interaction diagram
         :param progress_bar: If set to True, displays the progress bar
 
-        :raises ValueError: If ``control_points``, ``labels`` or ``n_points`` is invalid
-
         :return: Moment interaction results object
         """
 
-        # if an integer is provided for n_points, generate a list
-        if isinstance(n_points, int):
-            n_points = [n_points] * (len(control_points) - 1)
+        # compute extreme tensile fibre
+        _, d_t = utils.calculate_extreme_fibre(
+            points=self.compound_geometry.points, theta=theta
+        )
 
-        # if there are no labels provided, generate a list
-        if len(labels) == 1:
-            labels = labels * len(control_points)
+        # validate limits length
+        if len(limits) != 2:
+            raise ValueError("Length of limits must equal 2.")
 
-        # validate n_points length
-        if len(n_points) != len(control_points) - 1:
-            raise ValueError(
-                "Length of n_points must be one less than the length of control_points."
-            )
+        # get neutral axis depths for limits
+        limits_dn = []
+
+        for cp in limits:
+            limits_dn.append(self.decode_d_n(theta=theta, cp=cp, d_t=d_t))
 
-        # validate n_points entries are all longer than 1
-        for n_pt in n_points:
-            if n_pt < 2:
-                raise ValueError("n_points entries must be greater than 1.")
+        # get neutral axis depths for additional control points
+        add_cp_dn = []
+
+        for cp in control_points:
+            add_cp_dn.append(self.decode_d_n(theta=theta, cp=cp, d_t=d_t))
 
         # validate labels length
-        if len(labels) != len(control_points):
-            raise ValueError("Length of labels must equal length of control_points.")
+        if labels and len(labels) != 1 and len(labels) != 2 + len(control_points):
```

Comment:
should this be with an or as implied in the valueerror text?:-
```suggestion
        if labels and (len(labels) != 1 or len(labels) != 2 + len(control_points)):
```

Reply:
Actually, I think your logic is correct here. Was reading it as equal to!